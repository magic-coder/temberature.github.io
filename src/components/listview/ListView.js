import React from "react";
import "./ListView.less";
import ReactDOM from "react-dom";

class ListView extends React.Component {
  constructor() {
    super();
    this.state = {
      items: [1, 2, 3]
    };
    this.handleScroll();
  }
  handleScroll() {
    var locked = false, batch = this.state.items.length;
    document.addEventListener("scroll", () => {
      console.log(this.refs);
      if (locked) {
        return;
      }
      var hit = this.hitBottomTest(ReactDOM.findDOMNode(this.refs.container));
      console.log(hit);
      if (hit) {
        locked = true;
        this.setState(prev => {
          var is = prev.items,
            l = is.length;
          for (var i = 0; i < batch; i++) {
            this.state.items[l + i] = is[i] + l;
            console.log(this.state.items);
          }
          locked = false;
        });
      }
    });
  }
  hitBottomTest(elem) {
    var screenHeight = window.innerHeight;
    var scrollTop = document.scrollTop;
    // return elementBottom - screenHeight - scrollTop >= 0;
    return elem.getBoundingClientRect().bottom <= screenHeight;
  }
  render() {
    return (
      <div className="listview" ref="container">
        {this.state.items.map(val => (
          <div className="row" key={val}>
            {val}
          </div>
        ))}
      </div>
    );
  }
}
export default ListView;
