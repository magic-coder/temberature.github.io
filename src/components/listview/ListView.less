.listview {
  .row {
    // height: 200px;
    // border: 1px solid black;
    // font-size: 30px;
    // text-align: center;
  }
}
