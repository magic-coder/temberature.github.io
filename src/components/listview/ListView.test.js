import React from "react";
import renderer from "react-test-renderer";
import ListView from "./ListView";

it("renders correctly", function() {
  const data = [{
    "course_id": 1,
    "title": "认识电影：电影制作与电影艺术",
    "address": "北京市朝阳区国贸建外SOHO东区2号楼ATA",
    "event_start_date": 1509760800000,
    "event_end_date": 1534464000000,
    "event_register_deadline": 1510318800000,
    "latitude": 39.905624,
    "longitude": 116.459945,
    "textbook_title": "认识电影",
    "max_attendence": 30,
    "enrolled_already": 1,
    "event_main_content": "赘言絮语\n\n但凡喜欢观看电影，同时也喜欢阅读的朋友，家里都会藏有两本书，一本是贾内梯的《认识电影》，另外一本是大卫波德维尔的《电影艺术》。可能大多数人都没有机会仔细阅读这两本书，但，放在书架上，也有一种坐拥宝库的单纯的幸福。不过心里难免是有隐约遗憾的：每年观看近百部电影，如果我能更深入的理解电影艺术，那该多好！\n\n机会终于来了。在广大OA伙伴们的鼎沸呼吁下，奥林浦斯学院在这一期新课里，终于加入了这门久经期待的《认识电影：电影艺术与电影制作》的沙龙活动。\n\n简介：\n这门课，将以贾内梯的《认识电影》为核心教材，同时以大卫波德维尔的《电影艺术》为辅助阅读材料。在为期八周的课程里，你将学会从摄影、剪辑、叙事、表演以及电影理论等不同角度来解构与欣赏电影，在课程结束时，你将会与同学结成小组，拍摄一部3-5分钟的电影短篇，作为你的结课作品。\n\n通过本活动，你将收获什么？\n1. 学会从各种角度欣赏电影艺术，如摄影构图、灯光与气氛塑造、连续剪辑与空间蒙太奇、摄影机的运动方式、电影音乐的选择、电影理论里的结构主义和后结构主义等；\n2. 在两个月的时间里用解构的新颖方式观看大约30部经典电影；\n3. 将学到的知识运用于实践，尝试拍摄人生的第一部电影短片。\n\n安排：\n时长：8周\n时间：2017年11月4日 ~ 2017年12月30日 每周六上午10:00-12:30。11月4日为见面会。第1次正式课为11月11日，第8次（最后一次）为12月30日。\n地点：朝阳国贸\n名额：30人\n\n教材：\n1. 核心教材：【美】 路易斯·贾内梯. 2007.《认识电影》. 第11版. 北京. 世界图书出版公司\n豆瓣介绍：https://book.douban.com/subject/2326403/ 9096人评价，评分：8.9\n《认识电影》一书诞生于20世纪70年代，历经12次修订，是当今世界发行量、流传度皆高居榜首的电影入门书。经四十余年的洗礼，此书已成为百余所世界名校电影概论、影视制作、电影美学等专业的经典教材和必读书目。\n\n2. 辅助教材：【美】大卫·波德维尔. 2008. 《电影艺术：形式与风格》. 第8版. 北京：世界图书出版公司\n豆瓣介绍：https://book.douban.com/subject/3227478/ 2191人评价，评分：8.9\n作者以一个非常系统性的方式进行电影“形式”与“风格”方面的概念思考，同时佐以巨细靡遗的影片举例，详细分析电影的元素，说明电影的形式系统（叙事性与非叙事性）及风格系统（摄影、剪辑、场面调度与声音）如何在影片中交互作用。这样能帮助读者在了解一部电影时，不再是以过往观影经验所累积的惯性直觉方式，去感受影片技术上或情感上的精彩片段，而能将电影看作如同诗歌、建筑、音乐与舞蹈等其他艺术一般，是一个创作者在凝粹创作意念之后，执行到影片胶卷上的完整呈现。",
    "event_frontcover_filepath": "/25/39/cf/d5/2539cfd555031a69c4a572a8d01dbf56.jpg"
  }];
  const tree = renderer
    .create(
      <ListView
      dataSource={data}
      renderRow={(item) => <div key={item.course_id}>row</div>}
      renderHeader={() => <div>header</div>}
      renderSeparator={(item) => <div key={item.course_id * 2}>separator</div>}
      renderFooter={() => <div>footer</div>}
      useBodyScroll
      fixed={false}
    />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
