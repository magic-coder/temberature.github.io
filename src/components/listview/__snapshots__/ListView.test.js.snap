// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="oa-list-view-scrollview"
>
  <div
    className="oa-list-header"
    style={
      Object {
        "position": "",
        "top": 0,
      }
    }
  >
    <div>
      header
    </div>
  </div>
  <div
    className="oa-list-view-section-body"
  >
    <div>
      row
    </div>
    <div>
      separator
    </div>
  </div>
  <div
    className="oa-list-footer"
  >
    <div>
      footer
    </div>
  </div>
</div>
`;
